### PR TITLE
Display background when printing or saving a text widget (issue #14928)

### LINKED ISSUE #14928: [annotation] Text widget background color is not taken into account when printed
Attach (recommended) or Link to PDF file here:
[cerfa_15615-01-2.pdf](https://github.com/mozilla/pdf.js/files/8710789/cerfa_15615-01-2.pdf)

Configuration:
- Web browser and its version: Firefox nightly
- Operating system and its version: Windows 11

Steps to reproduce the problem:
1. File the 3rd field from the top-left on page 1 (i.e. `Né(e) le`): 
image

2. print

The printed field is unreadable:
image

The annotation has a white background which should be used when the appearance is generated when printing.

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

**Review by Snuffleupagus — COMMENTED** [review]
The shorter number format in the written data, especially for integer values, is a nice bonus of these changes!

File: src/core/annotation.js [hunk 1]
```diff
@@ -1640,10 +1673,12 @@ class WidgetAnnotation extends Annotation {
       );
     }
 
+    const colors = this.getBorderAndBackgroundAppearance();
+
     if (alignment === 0 || alignment > 2) {
       // Left alignment: nothing to do
       return (
-        "/Tx BMC q BT " +
+        `/Tx BMC q ${colors}BT ` +
```
[the remark this answers is not in the document]
Comment (calixteman) [reply src/core/annotation.js:1681]: It was to avoid to have two spaces in case it's empty.

[the remark this answers is not in the document]
Comment (Snuffleupagus) [reply src/core/annotation.js]: That feels a bit unfortunate, from a readability point of view, but hopefully the existing unit-tests would catch any breakage here; hence I suppose that we can leave this as-is then.

[the remark this answers is not in the document]
Comment (calixteman) [reply src/core/annotation.js:1681]: I wasn't happy myself with either the double space or the non-space between colors and BT...
But I always complain myself when I see some useless spaces in pdfs streams...
So it's why I opted for this solution.
I thought to have an array, push the commands, filter out empty ones and then join but I thought it was a bit overkilled.
Another solution could be to have `q ${colors ? colors + " " : ""}BT` but I'm not sure it's improving readability.

**Comment by Snuffleupagus** [comment]
(It seems that Chrome timed-out on the Linux bot...)

Are all of these changes consistent with printing from e.g. Adobe Reader?
Looking at `160F-2019-page1`, that one doesn't seem to include any border for that field when printing from Adobe Reader (although that particular field isn't editable there).

**Comment by calixteman** [comment]
in `160F-2019.pdf`, the colors of background and border are set thanks to JS in a `WillPrint` action and reset in `DidPrint` one.
But it won't work correctly because of:
https://github.com/mozilla/pdf.js/blob/3c877f9520006ab302a867790e18b0167444c23f/web/app.js#L1773-L1774

Anyway I began to add the color values in annotationStorage when they're modified by JS. If you're ok I'll make these changes in an other PR.
If you've a good idea to be able to wait for `WillPrint` results, please propose.

**Comment by calixteman** [comment]
> in `160F-2019.pdf`, the colors of background and border are set thanks to JS in a `WillPrint` action and reset in `DidPrint` one. But it won't work correctly because of:
> 
> https://github.com/mozilla/pdf.js/blob/3c877f9520006ab302a867790e18b0167444c23f/web/app.js#L1773-L1774
> 
> Anyway I began to add the color values in annotationStorage when they're modified by JS. If you're ok I'll make these changes in an other PR. If you've a good idea to be able to wait for `WillPrint` results, please propose.

I'll try to fix that in a follow-up.

**Comment by Snuffleupagus** [comment]
> Anyway I began to add the color values in annotationStorage when they're modified by JS. If you're ok I'll make these changes in an other PR.

Yeah, this one is probably large enough already so that sounds good to me.

> If you've a good idea to be able to wait for `WillPrint` results, please propose.

Unfortunately not, since it seems to be a web-platform limitation given that printing is synchronous.
It seems that you'd really want *async* printing support (which as far as I understand doesn't exist on the web), since trying to somehow delay regular printing seems like it could easily cause lots of issues.

**Pull request merged**